cluster-controller sending on closed channel

The cluster controller at syseleven just crashed with one of its goroutines and is dandling in an inconsistent state, not terminating. 
I only found out because Prometheus alerts where sent into your #alerting-syseleven slack channel and I looked into the alert.

```
I0208 07:11:42.222519       8 leaderelection.go:243] lock is held by cluster-controller-v1-157767814-6g22f-kubermatic-cluster-controller and has not yet expired
I0208 07:11:42.238889       8 leaderelection.go:203] failed to renew lease kube-system/kubermatic-cluster-controller: timed out waiting for the condition
I0208 07:11:42.268194       8 main.go:196] Event(v1.ObjectReference{Kind:"Endpoints", Namespace:"kube-system", Name:"kubermatic-cluster-controller", UID:"45fed553-e410-11e7-aa6b-fa163e12ceaa", APIVersion:"v1", ResourceVersion:"90690875", FieldPath:""}): type: 'Normal' reason: 'LeaderElection' cluster-controller-v1-157767814-gb9rd-kubermatic-cluster-controller stopped leading
E0208 07:11:42.268272       8 main.go:162] ==================== OnStoppedLeading ====================
I0208 07:11:42.519902       8 controller.go:371] Shutting down cluster controller
E0208 07:11:42.519947       8 main.go:113] Stopping os signal notifier: <nil>
E0208 07:11:42.519954       8 main.go:136] Stopping internal http server: <nil>
I0208 07:11:42.522830       8 main.go:140] Shutting down the internal http server
E0208 07:11:42.732666       8 main.go:180] Stopping controller: <nil>
panic: close of closed channel

goroutine 1 [running]:
main.main.func6(0x0, 0x0)
	/go/src/github.com/kubermatic/kubermatic/api/cmd/kubermatic-cluster-controller/main.go:181 +0x9c
github.com/kubermatic/kubermatic/api/vendor/github.com/oklog/run.(*Group).Run(0xc420549e30, 0xc4201aac00, 0x2)
	/go/src/github.com/kubermatic/kubermatic/api/vendor/github.com/oklog/run/group.go:47 +0x141
main.main()
	/go/src/github.com/kubermatic/kubermatic/api/cmd/kubermatic-cluster-controller/main.go:185 +0xbaf
```

Sadly the controller keeps on running as it is in a inconsistent state.
Like this it seems that it is able to stay in a deadlock or something, but it's not possible to be scraped by Prometheus in this state anymore.
It is just sitting there doing nothing and blocks new instances being scheduled.

@mrIncompetent

--- a/api/cmd/kubermatic-cluster-controller/main.go
+++ b/api/cmd/kubermatic-cluster-controller/main.go
@@ -2,12 +2,10 @@ package main
 
 import (
 	"context"
+	"errors"
 	"flag"
 	"fmt"
 	"net/http"
-	"os"
-	"os/signal"
-	"syscall"
 	"time"
 
 	"github.com/golang/glog"
@@ -21,6 +19,7 @@ import (
 	"github.com/kubermatic/kubermatic/api/pkg/provider"
 	"github.com/kubermatic/kubermatic/api/pkg/provider/cloud"
 	"github.com/kubermatic/kubermatic/api/pkg/provider/seed"
+	"github.com/kubermatic/kubermatic/api/pkg/signals"
 	"github.com/oklog/run"
 	"github.com/prometheus/client_golang/prometheus/promhttp"
 
@@ -101,17 +100,20 @@ func main() {
 	// Create event recorder
 	recorder := getEventRecorder(masterKubeClient)
 
+	stopCh := signals.SetupSignalHandler()
+	ctx, ctxDone := context.WithCancel(context.Background())
+
 	// This group is forever waiting in a goroutine for signals to stop
 	{
-		sig := make(chan os.Signal, 2)
 		g.Add(func() error {
-			signal.Notify(sig, os.Interrupt, syscall.SIGTERM)
-			glog.Info("Waiting for signal to stop")
-			<-sig
-			return nil
+			select {
+			case <-stopCh:
+				return errors.New("user requested to stop the application")
+			case <-ctx.Done():
+				return errors.New("parent context has been closed - propagating the request")
+			}
 		}, func(err error) {
-			glog.Errorf("Stopping os signal notifier: %v", err)
-			close(sig)
+			ctxDone()
 		})
 	}
 
@@ -136,7 +138,7 @@ func main() {
 			return nil
 		}, func(err error) {
 			glog.Errorf("Stopping internal http server: %v", err)
-			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+			ctx, cancel := context.WithTimeout(ctx, time.Second)
 			defer cancel()
 
 			glog.Info("Shutting down the internal http server")
@@ -148,21 +150,22 @@ func main() {
 
 	// This group is running the actual controller logic
 	{
-		leaderElectionStop := make(chan struct{})
-		controllerStop := make(chan struct{})
 		g.Add(func() error {
-			leaderElectionClient := kubernetes.NewForConfigOrDie(restclient.AddUserAgent(masterConfig, "kubermatic-cluster-controller-leader-election"))
+			leaderElectionClient, err := kubernetes.NewForConfig(restclient.AddUserAgent(masterConfig, "kubermatic-cluster-controller-leader-election"))
+			if err != nil {
+				return err
+			}
 			callbacks := kubeleaderelection.LeaderCallbacks{
 				OnStartedLeading: func(stop <-chan struct{}) {
-					err := startController(controllerStop, dcs, masterConfig, seedCmdConfig, versions, updates)
+					err := startController(ctx.Done(), dcs, masterConfig, seedCmdConfig, versions, updates)
 					if err != nil {
 						glog.Error(err)
+						ctxDone()
 					}
-					close(leaderElectionStop)
 				},
 				OnStoppedLeading: func() {
 					glog.Error("==================== OnStoppedLeading ====================")
-					close(controllerStop)
+					ctxDone()
 				},
 			}
 
@@ -176,11 +179,11 @@ func main() {
 			}
 
 			go leader.Run()
-			<-leaderElectionStop
+			<-ctx.Done()
 			return nil
 		}, func(err error) {
 			glog.Errorf("Stopping controller: %v", err)
-			close(controllerStop)
+			ctxDone()
 		})
 	}
 
@@ -201,7 +204,7 @@ func getEventRecorder(masterKubeClient *kubernetes.Clientset) record.EventRecord
 	return recorder
 }
 
-func startController(stop chan struct{}, dcs map[string]provider.DatacenterMeta, masterConfig *restclient.Config, seedConfig *clientcmdapi.Config, versions map[string]*apiv1.MasterVersion, updates []apiv1.MasterUpdate) error {
+func startController(stop <-chan struct{}, dcs map[string]provider.DatacenterMeta, masterConfig *restclient.Config, seedConfig *clientcmdapi.Config, versions map[string]*apiv1.MasterVersion, updates []apiv1.MasterUpdate) error {
 	metrics := NewClusterControllerMetrics()
 	clusterMetrics := cluster.ControllerMetrics{
 		Clusters: metrics.Clusters,

--- a/api/pkg/controller/cluster/controller.go
+++ b/api/pkg/controller/cluster/controller.go
@@ -41,7 +41,7 @@ const (
 // GroupRunStopper represents a control loop started with Run,
 // which can be terminated by closing the stop channel
 type GroupRunStopper interface {
-	Run(workerCount int, stop chan struct{})
+	Run(workerCount int, stop <-chan struct{})
 }
 
 // SeedClientProvider offers functions to get resources of a seed-cluster
@@ -304,7 +304,7 @@ func (cc *controller) syncInPhase(phase kubermaticv1.ClusterPhase) {
 	}
 }
 
-func (cc *controller) Run(workerCount int, stopCh chan struct{}) {
+func (cc *controller) Run(workerCount int, stopCh <-chan struct{}) {
 	defer utilruntime.HandleCrash()
 
 	cc.metrics.Workers.Set(float64(workerCount))

--- a/api/pkg/signals/signal.go
+++ b/api/pkg/signals/signal.go
@@ -1,0 +1,31 @@
+package signals
+
+import (
+	"os"
+	"os/signal"
+	"syscall"
+)
+
+var (
+	onlyOneSignalHandler = make(chan struct{})
+	shutdownSignals      = []os.Signal{os.Interrupt, syscall.SIGTERM}
+)
+
+// SetupSignalHandler registered for SIGTERM and SIGINT. A stop channel is returned
+// which is closed on one of these signals. If a second signal is caught, the program
+// is terminated with exit code 1.
+func SetupSignalHandler() (stopCh <-chan struct{}) {
+	close(onlyOneSignalHandler) // panics when called twice
+
+	stop := make(chan struct{})
+	c := make(chan os.Signal, 2)
+	signal.Notify(c, shutdownSignals...)
+	go func() {
+		<-c
+		close(stop)
+		<-c
+		os.Exit(1) // second signal. Exit directly.
+	}()
+
+	return stop
+}

